northstar/src/runtime/containerdb.rs

```diff
@@ -0,0 +1,211 @@
+//! Defines [ContainerDB], a struct to manage containers
+//!
+//! Northstar organizes containers in repositories. Repositories can be any object that implements
+//! the [Repository][crate::runtime::repository::Repository] trait.
+
+use std::{collections::HashMap, ops::Deref, sync::Arc};
+
+use bytes::Bytes;
+use tokio::sync::mpsc::Receiver;
+
+use crate::{
+    common::container::Container,
+    npk::manifest::Manifest,
+    runtime::{
+        error::Error,
+        repository::{Repository, RepositoryId},
+        state::Npk,
+    },
+};
+
+/// Container Database
+#[derive(Debug, Default)]
+pub struct ContainerDB {
+    index_by_repository: HashMap<RepositoryId, usize>,
+    index_by_container: HashMap<Container, usize>,
+    repositories: Vec<(RepositoryId, Box<dyn Repository>)>,
+}
+
+impl ContainerDB {
+    /// Insert a new container repository into the database
+    pub fn insert<R>(&mut self, id: RepositoryId, repository: R) -> Result<(), Error>
+    where
+        R: Repository + 'static,
+    {
+        // Check for repository index collition
+        if self.index_by_repository.contains_key(&id) {
+            return Err(Error::InvalidRepository(id));
+        }
+
+        // Check for container index collitions
+        for container in repository.list() {
+            if self.index_by_container.contains_key(&container) {
+                return Err(Error::DuplicateContainer(container));
+            }
+        }
+
+        // make the insertion
+        let index = self.repositories.len();
+
+        self.index_by_repository.insert(id.clone(), index);
+        for container in repository.list() {
+            self.index_by_container.insert(container, index);
+        }
+
+        self.repositories.push((id, Box::new(repository)));
+        Ok(())
+    }
+
+    /// Find the container's NPK
+    pub fn find_npk(&self, container: &Container) -> Option<Arc<Npk>> {
+        self.find_repository(container)
+            .and_then(|repository| repository.get(container))
+    }
+
+    /// Find the container's manifest
+    pub fn find_manifest(&self, container: &Container) -> Option<impl Deref<Target = Manifest>> {
+        struct ManifestRef(Arc<Npk>);
+
+        impl Deref for ManifestRef {
+            type Target = Manifest;
+
+            fn deref(&self) -> &Self::Target {
+                self.0.manifest()
+            }
+        }
+
+        self.find_npk(container).map(ManifestRef)
+    }
+
+    /// Find the repository that has the given container
+    pub fn find_repository(&self, container: &Container) -> Option<&dyn Repository> {
+        self.index_by_container
+            .get(container)
+            .and_then(|&idx| self.repositories.get(idx).map(|(_, b)| b.as_ref()))
```

Comment:
Index operations scare me...

Reply:
i'd agree, on the other hand this is not unsafe, should not panic if there is a bug,